File: files/routes/casino.py
import json
from files.__main__ import app
from files.helpers.wrappers import *
from files.helpers.alerts import *
from files.helpers.get import *
from files.helpers.const import *
from files.helpers.wrappers import *
from files.helpers.blackjack import *
from files.helpers.slots import *
from files.helpers.lottery import *


@app.get("/casino")
@auth_required
def casino(v):
    participants = get_users_participating_in_lottery()
    return render_template("casino.html", v=v, participants=participants)


@app.post("/casino/slots")
@auth_required
def pull_slots(v):
    try:
        wager = int(request.values.get("wager"))
    except:
        return {"error": "Invalid wager."}

    try:
        currency = request.values.get("currency")
    except:
        return {"error": "Invalid currency (expected 'dramacoin' or 'marseybux')."}

    if (currency == "dramacoin" and wager > v.coins) or (currency == "marseybux" and wager > v.procoins):
        return {"error": f"Not enough {currency} to make that bet."}

    success, game_state = casino_slot_pull(v, wager, currency)

    if success:
        return {"game_state": game_state}
    else:
        return {"error": "Wager must be more than 100 {currency}."}

@app.get("/casino/blackjack")
@auth_required
def get_player_blackjack_status(v):
    game_state = get_safe_game_state(v)

    if game_state is not None:
        return {"active": True, "game_state": game_state}
    else:
        return {"active": False}

@app.post("/casino/blackjack")
@auth_required
def deal_blackjack(v):
    try:
        wager = int(request.values.get("wager"))
    except:
        return {"error": "Invalid wager."}

    try:
        currency = request.values.get("currency")
    except:
        return {"error": "Invalid currency (expected 'dramacoin' or 'marseybux')."}

    if (currency == "dramacoin" and wager > v.coins) or (currency == "marseybux" and wager > v.procoins):
        return {"error": f"Not enough {currency} to make that bet."}

    game = get_active_game(v)

    if game:
        return {"error": "This play is already playing a game."}

    success = deal_blackjack_game(v, wager, currency)

    if success:
        game_state = get_safe_game_state(v)

        return {"game_state": game_state}
    else:
        return {"error": "Wager must be more than 100 {currency}."}


@app.post("/casino/blackjack/action")
@auth_required
def player_took_blackjack_action(v):
    try:
        action = request.values.get("action")
    except:
        return {"error": "Invalid action."}

    game_state = get_safe_game_state(v)

    if not game_state:
        return {"error": "No game exists for player."}

    elif action == 'hit':
        gambler_hit(v)
    elif action == 'stay':
        gambler_stayed(v)
    elif action == 'double_down':
        gambler_doubled_down(v)
    elif action == 'insure':
        gambler_purchased_insurance(v)

    game_state = get_safe_game_state(v)

    if game_state:
        return {"active": True, "game_state": game_state}
    else:
        return {"active": False}

File: files/assets/js/casino.js
// Slots
function pullSlots() {
  const wager = document.getElementById("casinoSlotsBet").value;
  const currency = document.querySelector(
    'input[name="casinoSlotsCurrency"]:checked'
  ).value;

  document.getElementById("casinoSlotsBet").disabled = true;
  document.getElementById("casinoSlotsPull").disabled = true;

  const xhr = new XMLHttpRequest();
  xhr.open("post", "/casino/slots");
  xhr.onload = handleSlotsResponse.bind(null, xhr);

  const form = new FormData();
  form.append("formkey", formkey());
  form.append("wager", wager);
  form.append("currency", currency);

  xhr.send(form);
}

function handleSlotsResponse(xhr) {
  let response;

  try {
    response = JSON.parse(xhr.response);
  } catch (error) {
    console.error(error);
  }

  const succeeded =
    xhr.status >= 200 && xhr.status < 300 && response && !response.error;
  const slotsResult = document.getElementById("casinoSlotsResult");
  slotsResult.classList.remove("text-success", "text-danger");

  if (succeeded) {
    const { game_state } = response;
    const state = JSON.parse(game_state);
    const reels = Array.from(document.querySelectorAll(".reel"));
    const symbols = state.symbols.split(",");

    for (let i = 0; i < 3; i++) {
      reels[i].innerHTML = symbols[i];
    }

    slotsResult.style.display = "block";
    slotsResult.innerText = state.text;

    if (state.text.includes("Won")) {
      if (state.text.includes("Jackpot")) {
        slotsResult.classList.add("text-warning");
      } else {
        slotsResult.classList.add("text-success");
      }
    } else if (state.text.includes("Lost")) {
      slotsResult.classList.add("text-danger");
    }
  } else {
    slotsResult.style.display = "block";
    slotsResult.innerText = response.error;
    slotsResult.classList.add("text-danger");

    console.error(response.error);
  }

  document.getElementById("casinoSlotsBet").disabled = false;
  document.getElementById("casinoSlotsPull").disabled = false;
}

// Blackjack
function handleBlackjackResponse(xhr) {
  let response;

  try {
    response = JSON.parse(xhr.response);
  } catch (error) {
    console.error(error);
  }

  const succeeded =
    xhr.status >= 200 && xhr.status < 300 && response && !response.error;
  const blackjackResult = document.getElementById("casinoBlackjackResult");
  blackjackResult.classList.remove("text-success", "text-danger");

  if (succeeded) {
    updateBlackjack(response.game_state);
  } else {
    blackjackResult.style.display = "block";
    blackjackResult.innerText = response.error;
    blackjackResult.classList.add("text-danger");

    console.error(response.error);
  }
}

function updateBlackjack(state) {
    const { player, dealer, status } = state;
    const lettersToSuits = {
      S: "♠️",
      H: "♥️",
      C: "♣️",
      D: "♦️",
    };
    const suitsToColors = {
      "♠️": "black",
      "♥️": "red",
      "♣️": "black",
      "♦️": "red",
    };

    // Clear everything.
    Array.from(document.querySelectorAll(".playing-card")).forEach((card) => {
      card.innerText = "";
      card.style.color = "unset";
      card.classList.remove("dealt");
    });

    // Show dealer cards.
    const dealerSlots = Array.from(
      document.querySelectorAll('.playing-card[data-who="dealer"]')
    );
    for (let i = 0; i < dealer.length; i++) {
      const slot = dealerSlots[i];

      if (slot) {
        // Technically, the dealer can use more than 5 cards, though it's rare.
        // In that case, the result message is good enough.
        // Thanks, Carp. 🐠
        slot.classList.add("dealt");

        if (i > 0 && status === "active") {
          break;
        }

        const rank = dealer[i][0];
        const suit = lettersToSuits[dealer[i][1]];
        const card = rank + suit;
        slot.innerText = card;
        slot.style.color = suitsToColors[suit];
      }
    }

    // Show player cards.
    const playerSlots = Array.from(
      document.querySelectorAll('.playing-card[data-who="player"]')
    );
    for (let i = 0; i < player.length; i++) {
      const slot = playerSlots[i];
      const rank = player[i][0];
      const suit = lettersToSuits[player[i][1]];
      const card = rank + suit;
      slot.innerText = card;
      slot.style.color = suitsToColors[suit];
      slot.classList.add("dealt");
    }

    updateBlackjackActions(state);
}

function buildBlackjackAction(id, method, title) {
  return `
    <button
      type="button"
      class="btn btn-secondary lottery-page--action"
      id="${id}"
      onclick="${method}()"
    >
      ${title}
    </button>
  `;
}

function updateBlackjackActions(state) {
  // Clear all actions first.
  const actionWrapper = document.getElementById("casinoBlackjackActions");
  actionWrapper.innerHTML = "";

  if (state.status === "active") {
    const actions = [];
    const hit = buildBlackjackAction(
      "casinoBlackjackHit",
      "hitBlackjack",
      "Hit"
    );
    const stay = buildBlackjackAction(
      "casinoBlackjackStay",
      "stayBlackjack",
      "Stay"
    );
    const double = buildBlackjackAction(
      "casinoBlackjackDouble",
      "doubleBlackjack",
      "Double Down"
    );

    actions.push(hit, stay, double);

    if (state.dealer[0][0] === "A" && !state.insured) {
      const insure = buildBlackjackAction(
        "casinoBlackjackInsure",
        "insureBlackjack",
        "Insure"
      );

      actions.push(insure);
    }

    actionWrapper.innerHTML = actions.join("\n");
  } else {
    // Game is over, deal a new game.
    document.getElementById("casinoBlackjackWager").style.display = "block";

    const deal = buildBlackjackAction(
      "casinoBlackjackDeal",
      "dealBlackjack",
      "Deal"
    );

    actionWrapper.innerHTML = deal;
  }
}

// == Blackjacktions
function dealBlackjack() {
  const wager = document.getElementById("casinoBlackjackBet").value;
  const currency = document.querySelector(
    'input[name="casinoBlackjackCurrency"]:checked'
  ).value;

  document.getElementById("casinoBlackjackBet").disabled = true;
  document.getElementById("casinoBlackjackDeal").disabled = true;
  document.getElementById("casinoBlackjackWager").style.display = "none";

  const xhr = new XMLHttpRequest();
  xhr.open("post", "/casino/blackjack");
  xhr.onload = handleBlackjackResponse.bind(null, xhr);

  const form = new FormData();
  form.append("formkey", formkey());
  form.append("wager", wager);
  form.append("currency", currency);

  xhr.send(form);
}

function takeBlackjackAction(action) {
  const xhr = new XMLHttpRequest();
  xhr.open("post", "/casino/blackjack/action");
  xhr.onload = handleBlackjackResponse.bind(null, xhr);

  const form = new FormData();
  form.append("formkey", formkey());
  form.append("action", action);

  xhr.send(form);
}

const hitBlackjack = takeBlackjackAction.bind(null, "hit");
const stayBlackjack = takeBlackjackAction.bind(null, "stay");
const doubleBlackjack = takeBlackjackAction.bind(null, "double");
const insureBlackjack = takeBlackjackAction.bind(null, "insure");

// When the casino loads, look up the "blackjack status" of a player to either start a new game or continue an existing game.
function checkBlackjackStatus() {
  const xhr = new XMLHttpRequest();
  xhr.open("get", "/casino/blackjack/1");
  xhr.onload = handleBlackjackStatusResponse.bind(null, xhr);
  xhr.send();
}

function handleBlackjackStatusResponse(xhr) {
  let response;

  try {
    response = JSON.parse(xhr.response);
  } catch (error) {
    console.error(error);
  }

  const succeeded =
    xhr.status >= 200 && xhr.status < 300 && response && !response.error;

  if (succeeded) {
    if (response.active) {
      updateBlackjack(response.game_state)
    }
  } else {
    console.error("error");
  }
}

if (
  document.readyState === "complete" ||
  (document.readyState !== "loading" && !document.documentElement.doScroll)
) {
  checkBlackjackStatus();
} else {
  document.addEventListener("DOMContentLoaded", checkBlackjackStatus);
}
